src/Databases/MySQL/DatabaseConnectionMySQL.cpp

```diff
@@ -60,11 +60,81 @@ static String toQueryStringWithQuote(const std::vector<String> & quote_list)
+        { std::make_shared<DataTypeUInt64>(),   "precision" },
+        { std::make_shared<DataTypeUInt64>(),   "scale" },
+    };
+
+    WriteBufferFromOwnString query;
+    query << "SELECT "
+             " TABLE_NAME AS table_name,"
+             " COLUMN_NAME AS column_name,"
+             " COLUMN_TYPE AS column_type,"
+             " IS_NULLABLE = 'YES' AS is_nullable,"
+             " COLUMN_TYPE LIKE '%unsigned' AS is_unsigned,"
+             " CHARACTER_MAXIMUM_LENGTH AS length,"
+             " NUMERIC_PRECISION as '',"
+             " IF(ISNULL(NUMERIC_SCALE), DATETIME_PRECISION, NUMERIC_SCALE) AS scale" // we know DATETIME_PRECISION as a scale in CH
+             " FROM INFORMATION_SCHEMA.COLUMNS"
+             " WHERE TABLE_SCHEMA = " << quote << database_name
+          << " AND TABLE_NAME IN " << toQueryStringWithQuote(tables_name) << " ORDER BY ORDINAL_POSITION";
+
+    MySQLBlockInputStream result(pool.get(), query.str(), tables_columns_sample_block, DEFAULT_BLOCK_SIZE);
+    while (Block block = result.read())
+    {
+        const auto & table_name_col = *block.getByPosition(0).column;
+        const auto & column_name_col = *block.getByPosition(1).column;
+        const auto & column_type_col = *block.getByPosition(2).column;
+        const auto & is_nullable_col = *block.getByPosition(3).column;
+        const auto & is_unsigned_col = *block.getByPosition(4).column;
+        const auto & char_max_length_col = *block.getByPosition(5).column;
+        const auto & precision_col = *block.getByPosition(6).column;
+        const auto & scale_col = *block.getByPosition(7).column;
+
+        size_t rows = block.rows();
+        for (size_t i = 0; i < rows; ++i)
+        {
+            String table_name = table_name_col[i].safeGet<String>();
+            tables_and_columns[table_name].emplace_back(
+                    column_name_col[i].safeGet<String>(),
+                    convertMySQLDataType(
+                            type_support,
+                            column_type_col[i].safeGet<String>(),
+                            external_table_functions_use_nulls && is_nullable_col[i].safeGet<UInt64>(),
+                            is_unsigned_col[i].safeGet<UInt64>(),
+                            char_max_length_col[i].safeGet<UInt64>(),
+                            precision_col[i].safeGet<UInt64>(),
+                            scale_col[i].safeGet<UInt64>()));
+        }
+    }
+    return tables_and_columns;
+}
+
+
 DatabaseConnectionMySQL::DatabaseConnectionMySQL(
-    const Context & global_context_, const String & database_name_, const String & metadata_path_,
+    const Context & query_context_, const String & database_name_, const String & metadata_path_,
     const ASTStorage * database_engine_define_, const String & database_name_in_mysql_, mysqlxx::Pool && pool)
     : IDatabase(database_name_)
-    , global_context(global_context_.getGlobalContext())
+    , query_context(query_context_.getQueryContext())
```

Comment:
Maybe raise a null pointer?

Reply:
Yep, this could throw an exception if `query context` wasn't set, but this is compulsory to propagate settings from query, 
like:
```
SELECT * FROM mysql(...) SETTINGS mysql_datatypes_support_level='decimal';
```

But I have tests for both cases (with and without extra settings) and they work.

Reply:
Have you tested restart server? and the query_context is hold by database, I mean it's not to be released and destroyed(if it is released at the end my `create database` query,  there will be a problem at the next query)?

Reply:
It's not only for server start/stop. We had several crashes with JOIN over MySQL table where query context is destroyed while storage is still active. Or vice versa.
It's possible that we start fetching MySQL table  as right table in JOIN then decide that stream from left table fails (i.e. throws an error) and destory the query. So storage should be able to die with destroyed query context.